op-node/rollup/derive/blob_data_source.go

```diff
@@ -0,0 +1,119 @@
+package derive
+
+import (
+	"context"
+	"fmt"
+	"github.com/ethereum-optimism/optimism/op-node/eth"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type BlobsSidecar struct {
+	BeaconBlockRoot common.Hash    `json:"beacon_block_root"`
+	BeaconBlockSlot common.Hash    `json:"beacon_block_slot"`
+	Blobs           []types.Blob   `json:"blobs"`
+	AggregatedProof types.KZGProof `json:"kzg_aggregated_proof"`
+}
+
+type BlobsSidecarFetcher interface {
+	FetchSidecar(ctx context.Context, slot uint64) (*BlobsSidecar, error)
+}
+
+type BlobdataSource struct {
+	log          log.Logger
+	cfg          *rollup.Config
+	txFetcher    L1TransactionFetcher
+	blobsFetcher BlobsSidecarFetcher
+}
+
+func NewBlobdataSource(log log.Logger, cfg *rollup.Config, txFetcher L1TransactionFetcher, blobsFetcher BlobsSidecarFetcher) *BlobdataSource {
+	return &BlobdataSource{log: log, cfg: cfg, txFetcher: txFetcher, blobsFetcher: blobsFetcher}
+}
+
+func (cs *BlobdataSource) OpenData(ctx context.Context, id eth.BlockID) (DataIter, error) {
+	info, txs, err := cs.txFetcher.InfoAndTxsByHash(ctx, id.Hash)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch transactions: %w", err)
+	}
+
+	// Based on the timestamp we check which slot is being used
+	l1GenesisTime := uint64(12434)
+	blockTime := uint64(12)
+	slot := (info.Time() - l1GenesisTime) / blockTime
+
+	sidecar, err := cs.blobsFetcher.FetchSidecar(ctx, slot)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch blobs sidecar for slot %d: %w", slot, err)
+	}
+	// compute kzg commitments and versioned hashes of each blob in the block
+	var computedHashes []common.Hash
+	for i, bl := range sidecar.Blobs {
+		commitment, ok := bl.ComputeCommitment()
+		if !ok {
+			return nil, fmt.Errorf("failed compute blob commitment for blob %d", i)
+		}
+		computedHashes = append(computedHashes, commitment.ComputeVersionedHash())
+	}
+	// check which blobs are sent into our inbox
+
+	var expectedHashes []common.Hash
+	for _, tx := range txs {
+		if tx.Type() == types.BlobTxType {
+			expectedHashes = append(expectedHashes, tx.BlobVersionedHashes()...)
+		}
+	}
+	if len(computedHashes) != len(expectedHashes) {
+		return nil, fmt.Errorf("got %d data hashes for beacon block %s (%d), but expected %d hashes from blob txs in execution block %s",
+			len(computedHashes), sidecar.BeaconBlockRoot, sidecar.BeaconBlockSlot, len(expectedHashes), id)
+	}
+	for i, h := range computedHashes {
+		if h != expectedHashes[i] {
+			return nil, fmt.Errorf("data hash mismatch %d (beacon block %s, slot %d): hash %d does not match expected hash from el block %s: %s",
```

Comment:
[question] does "el" in this context mean execution layer? If so, proposal to capitalize (EL) or replace with the more explicit 

```suggestion
			return nil, fmt.Errorf("data hash mismatch %d (beacon block %s, slot %d): hash %d does not match expected hash from execution block %s: %s",
```

motivation: conforms to the error message above:
```go
...  hashes from blob txs in execution block %s"
```